rqt_py_common/src/rqt_py_common/topic_helpers.py

```diff
@@ -70,106 +77,175 @@ def get_topic_names_and_types(node=None):
+
+    if type_name in ['float', 'float32', 'float64',
+                     'double', 'long double']:
         return float
 
-    elif type_name in ['string']:
-        # string constants are always stripped
+    # TODO(mlautman): char should be a string of length one. We do not currently support this
+    if type_name in ['char', 'wchar', 'string', 'wstring']:
         return str
 
-    elif type_name in [
+    if type_name in ['octet']:
+        return bytes
+
+    if type_name in [
             'int8', 'uint8',
             'int16', 'uint16',
             'int32', 'uint32',
-            'int64', 'uint64',
-            'char', 'byte']:
+            'int64', 'uint64']:
         return int
 
-    elif type_name in ["bool"]:
+    if type_name in ['bool', 'boolean']:
         return bool
 
-    else:
-        return None
+    return None
 
 
-def get_field_types(topic_name):
+def get_field_type(target, node=None):
     """
     Get the Python type of a specific field in the given registered topic.
 
-    If the field is an array, the type of the array's values are returned and the is_array flag is
-    set to True. This is a static type check, so it works for unpublished topics and with empty
-    arrays.
+    If the field is an array, the type of the array's values are returned and the is_array flag
+    is set to True
 
-    :param topic_name: name of field of a registered topic, ``str``, i.e. '/rosout/file'
+    :param target: name of field of a registered topic, ``str``, i.e. '/rosout/file'
     :returns: field_type, is_array
     """
-    # Note: Mlautman 11/2/18
-    #       In ROS2 multiple msg types can be used with a single topic making this
-    #       funciton a bad candidate to port to ROS2
-    logger = logging.get_logger('topic_helpers')
-    logger.error("get_field_type is not implemented in ROS2")
-    # get topic_type and message_evaluator
-    # topic_type, real_topic_name, _ = get_topic_type(topic_name)
-    # if topic_type is None:
-    #     # qDebug('topic_helpers.get_field_type(%s): get_topic_type failed' % (topic_name))
-    #     return None, False
-
-    # message_class = roslib.message.get_message_class(topic_type)
-    # if message_class is None:
-    #     qDebug('topic_helpers.get_field_type(%s): get_message_class(%s) failed' %
-    #            (topic_name, topic_type))
-    #     return None, False
-
-    # slot_path = topic_name[len(real_topic_name):]
-    # return get_slot_type(message_class, slot_path)
-    pass
+    topic_names_and_types = get_topic_names_and_types(node=node)
+    return _get_field_type(topic_names_and_types, target)
+
+
+def _get_field_type(topic_names_and_types, target):  # noqa: C901
+    """Testable helper function for get_field_type."""
+    logger = logging.get_logger('topic_helpers._get_field_type')
+    for name, types in topic_names_and_types:
+        # If the topic is a substring of the target param, then we have found a match
+        if target.find(name) >= 0:
```

Comment:
I don't understand why `target.find(name) > 0` would be a match? In the following code you expect `target == name` or use it as `target[len(name):]`.

Reply:
I changed this line of logic to `target.startswith(name)`

Reply:
The updated condition still seems to match unrelated. E.g. `target` being `/foobar/baz` and `name` being `/foo`.

Reply:
Good point. New commit covers that corner case and tests for it as well